backend/LexBoxApi/Services/SendReceiveService.cs

```diff
@@ -0,0 +1,89 @@
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using LexBoxApi.Config;
+// using LexCore.Entities;
+using LexData;
+// using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Options;
+using SIL.Progress;
+
+namespace LexBoxApi.Services;
+
+public class SendReceiveService
+{
+    private readonly LexBoxDbContext _dbContext;
+    private readonly IHgService _hgService;
+    private readonly IOptions<HgConfig> _hgOptions;
+    private readonly IOptions<SendReceiveConfig> _sendReceiveOptions;
+    private readonly IProgress _progress;
+
+    public SendReceiveService(LexBoxDbContext dbContext, IHgService hgService, IOptions<HgConfig> hgOptions, IOptions<SendReceiveConfig> sendReceiveOptions, IProgress progress)
+    {
+        _dbContext = dbContext;
+        _hgService = hgService;
+        _hgOptions = hgOptions;
+        _sendReceiveOptions = sendReceiveOptions;
+        _progress = progress;
+    }
+
+    public async Task<string> VerifyHgVersion()
+    {
+        string output = "";
+        using (Process hg = new()) {
+            string hgFilename = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "hg.exe" : "hg";
+            hg.StartInfo.FileName = System.IO.Path.Join("Mercurial", hgFilename);
+            hg.StartInfo.Arguments = "version";
+            hg.StartInfo.RedirectStandardOutput = true;
+
+            hg.Start();
+            output = hg.StandardOutput.ReadToEnd();
+            await hg.WaitForExitAsync();
+        }
+        return output;
+    }
+
+    public async Task<string> CloneProject(string projectCode, string destDir)
+    {
+        var repoUrl = $"{_hgOptions.Value.HgWebUrl}/hg/{projectCode}";
+        var flexBridgeOptions = new Dictionary<string, string>
+        {
+            { "fullPathToProject", destDir },
+            { "fdoDataModelVersion", _sendReceiveOptions.Value.FdoDataModelVersion },
+            { "languageDepotRepoName", "LexBox" },
+            { "languageDepotRepoUri", repoUrl },
+            { "deleteRepoIfNoSuchBranch", "false" },
+        };
+        LfMergeBridge.LfMergeBridge.Execute("Language_Forge_Clone", _progress, flexBridgeOptions, out string cloneResult);
+
+        if (_progress is StringBuilderProgress)
```

Comment:
you can do this in C# now. `if (_progress is StringBuilderProgress sbProgress)` and then use the `sbProgress` variable only inside the scope of the if statment, then you don't need to cast it inside the function multiple times.

Reply:
I knew you could do type guards with `is`, I just didn't know the syntax so I was still getting warnings about the `.Text` attribute not being present in `IProgress`. I'll tighten that up now that I know the shorthand syntax for assigning the cast to a new name.